### ISSUE TITLE
Store Layer class kwargs as instance attributes

### ISSUE THREAD

**Issue by ebenolson** [benanne/Lasagne#155]
By altering the Layer class a little:

	    def __init__(self, incoming, name=None, **kwargs):
		    for k in kwargs.keys():
		        self.__setattr__(k, kwargs[k])

		    .....

it would be possible to add new functionality that depends on per-layer hyperparameters without needing to alter the base code.

For example, per-layer weight decay could be implemented like this:

		layer1 = lasagne.layers.DenseLayer(
		    	input_layer, 
		    	num_units = 1,
		    	weight_decay = 1.0,
		)

		layer2 = lasagne.layers.DenseLayer(
		    	layer1, 
		    	num_units = 1,
		    	weight_decay = 0.1,
		)

		layer3 = lasagne.layers.DenseLayer(
		    	layer2, 
		    	num_units = 1,
		)

		all_layers = lasagne.layers.get_all_layers(layer3)

		l2_penalty = sum([lasagne.regulization.l2(layer) * get_attr(layer, 'weight_decay', 1) 
						  for layer in all_layers])

I think this would make it easier to avoid errors than declaring per-layer values in a separate part of the code, and by storing the values in the relevant Layer, it avoids needing to maintain a mapping between a separate container and the layers.

**Comment by craffel** [comment]
You actually don't need any additional code to do this; any class instance can have attributes added on the fly.  So, you could do the same as your example, e.g. by setting
```
layer1 = lasagne.layers.DenseLayer(input_layer, num_units=1)
layer1.weight_decay = 1.0
```
after which point `getattr(layer1, 'weight_decay')` or `layer1.weight_decay` would return `1.0`.  Also, it's worth noting that `lasagne.regularization.l2` currently calls `get_all_params`, not `get_params`, so calling e.g. `lasagne.regularization.l2(layer3)` would return regularization terms for all parameters in all layers in your example, not regularization terms for `layer3` only as it looks like your code intends to do.  This would be fixed *if the regularizers taking expressions rather than layers as input* by calling `lasagne.regularization.l2(layer3.get_params())` instead.

**Comment by ebenolson** [comment]
That's true, although I don't think it looks as clean, and prevents declaring layers compositionally.

Thanks for the tip on `l2`, that's been tripping me up for a while.

**Comment by benanne** [comment]
I don't think adding all kwargs as attributes is a good idea, because it makes it much easier to make mistakes that fail silently. If you create an instance of a `Layer` subclass that has arguments with default values that you wish to override, and you have a typo in the name of one of those arguments, everything will just work and the layer will use the default value. An exception should be thrown instead.

I know it's more verbose, but if you want to do something like this, I think @craffel's approach is the "correct" way to do this.

> This would be fixed *if the regularizers taking expressions rather than layers as input* by calling `lasagne.regularization.l2(layer3.get_params())` instead.

I agree. As I've mentioned many times before, the `regularization` module was an afterthought and needs to be revamped.

**Comment by f0k** [comment]
> I don't think adding all kwargs as attributes is a good idea, because it makes it much easier to make mistakes that fail silently.

+1! (We have the same problem in `get_output`, but I think we should keep methods accepting any `**kwargs` to a minimum.)

**Comment by ebenolson** [comment]
> I don't think adding all kwargs as attributes is a good idea, because it makes it much easier to make mistakes that fail silently.

That's a good point which I hadn't considered, but it could be fixed by adding:

		__allowed_kwargs__ = []

		class Layer(object):
		    def __init__(self, incoming, name=None, **kwargs):
		        for k in kwargs.keys():
		            if k in __allowed_kwargs__:
		                self.__setattr__(k, kwargs[k])            
		            else: 
		                raise AttributeError('Unknown kwarg {0}'.format(k))

Then in e.g. `regularization.py`

		layers.base.__allowed_kwargs__ += ['weight_decay',]

**Comment by benanne** [comment]
That seems like a lot of added complexity just to make `Layer` instances behave like dictionaries. Besides, what if I add a custom submodule that adds an allowed kwargs `wight_decay`? I want to be able to decay my wights ;)

Okay, that was a contrived example, but I don't think this is a good solution for this problem (or good software design).

**Comment by f0k** [comment]
> it could be fixed by adding [...]

See my proposed alternative above! You may not have seen it, given that your post came only 2 seconds after mine :)

**Comment by ebenolson** [comment]
Yeah, actually I do like that, it emphasizes that you're adding a non-default attribute to the layer.
